rust/macros/vtable.rs

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use proc_macro::{Delimiter, Group, TokenStream, TokenTree};
+use std::collections::HashSet;
+use std::fmt::Write;
+
+pub(crate) fn vtable(_attr: TokenStream, ts: TokenStream) -> TokenStream {
+    let mut it = ts.into_iter();
+
+    let mut tokens = Vec::new();
+
+    // Scan for the `trait` or `impl` keyword
+    let is_trait = loop {
+        let token = it.next().expect("unexpected end");
+        let keyword = match &token {
+            TokenTree::Ident(ident) => match ident.to_string().as_str() {
+                "trait" => Some(true),
+                "impl" => Some(false),
+                _ => None,
+            },
+            _ => None,
+        };
+        tokens.push(token);
+        if let Some(v) = keyword {
+            break v;
+        }
+    };
+
+    // Scan for the main body
+    // FIXME: `{}` is allowed within type as well, e.g. `impl Foo for Bar<{0}>`.
```

Comment:
Isn't the body always the last token? If so, I think it's better to use this fact to find the body, which would also address the case you mention here.